Fix KVStore client transaction size

thanethomson opened the pull request:
Fixes #46 

Also introduces some new Prometheus metrics and aggregate stats that track overall transaction throughput and throughput rate in bytes and bytes/sec (respectively). This aids with testing and integration testing of this improvement.

adizere commented on pkg/loadtest/client_kvstore.go:
```diff
@@ -1,14 +1,29 @@
+	"github.com/tendermint/tendermint/libs/common"
+)
+
+const minKVStoreTxSize int = 40
```
Not clear why 40 should be the minimum. Maybe add a comment explaining the rationale for this number, similar to comments inside function `ValidateConfig`.

thanethomson replied:
The broader question here is: what strategy do we use when constructing key/value pairs for the KVStore client?

Right now I'm trying to use as random an approach as possible to reduce chances of key collisions (to create unique transactions). Another approach would be a totally controlled one to ensure uniqueness, where each key is a combination of a client ID and a counter. In the latter case, the minimum transaction size must cater for a full load test with no duplicate transactions across all clients.

Unless one wants duplicate transactions?

adizere replied:
I think the current strategy is perfect for our purposes.

I can imagine a scenario where having duplicate transactions (accessing the same key) would be a useful feature, namely to test how contention affects performance. Other workload frameworks (e.g., YCSB) generate keys drawn from a specific distribution, such as uniform or zipfian, and offer thus the ability to create contention on the same key.

Such a feature, however, would be _useful for testing application-level performance primarily_, and would have no bearing on the consensus layer performance. For instance, high-contention workloads (zipfian) could test how efficient is the key-value store at handling high-contention access. Since we're currently only looking to investigate the consensus layer, maybe we can skip this feature and keep it in mind for later.

thanethomson replied:
`tm-load-test` allows you to build different clients with different kinds of behaviours, so it should be easy to eventually build an array of different clients to simulate different workloads.

adizere commented on pkg/loadtest/client_kvstore_test.go:
```diff
@@ -0,0 +1,66 @@
+		{loadtest.Config{Size: 30}, true},  // tx size is too small
+		{loadtest.Config{Size: 40}, false}, // tx size is just right
+		{loadtest.Config{Size: 100}, false},
+		{loadtest.Config{Size: 250}, false},
```
As a side note, in some experiments I plan to increase the tx size up to even 10KB.

thanethomson replied:
Will add some much larger ones in! 👍

thanethomson commented:
I've modified the way I generate transactions in [this commit](https://github.com/interchainio/tm-load-test/pull/47/commits/d893746cea1b113eb61490874b4f6e50556f8ecb) now (different to the initial mechanism in the PR). Also added some benchmarks for tx generation just as an added sanity check. As expected, larger tx sizes take significantly longer to generate:

```
BenchmarkKVStoreClient_GenerateTx_32b-8     	 3061263	       387 ns/op
BenchmarkKVStoreClient_GenerateTx_64b-8     	 1986664	       604 ns/op
BenchmarkKVStoreClient_GenerateTx_128b-8    	 1291180	       930 ns/op
BenchmarkKVStoreClient_GenerateTx_256b-8    	  784158	      1550 ns/op
BenchmarkKVStoreClient_GenerateTx_512b-8    	  419467	      2735 ns/op
BenchmarkKVStoreClient_GenerateTx_1kB-8     	  231030	      4950 ns/op
BenchmarkKVStoreClient_GenerateTx_10kB-8    	   23889	     49719 ns/op
BenchmarkKVStoreClient_GenerateTx_100kB-8   	    2360	    469271 ns/op
```

Benchmark run on my MacBook Pro 2015. Middle column is number of txs generated in 1 second by a single client. So far this seems like, for small transaction sizes (< 10kB) the client itself should not impose any bottleneck in terms of Tendermint network throughput.
